File: Week 23/advanced_dp.rs
//! # Week 23: Advanced Dynamic Programming
//!
//! This module covers advanced DP techniques including bitmask DP, optimal
//! binary search, matrix chain multiplication, and digit DP.
//!
//! ## Complexity Summary
//! | Problem                     | Time            | Space         |
//! |----------------------------|-----------------|---------------|
//! | TSP (Bitmask DP)           | O(n^2 * 2^n)   | O(n * 2^n)    |
//! | LIS (O(n log n))           | O(n log n)      | O(n)          |
//! | Matrix Chain Multiplication| O(n^3)          | O(n^2)        |
//! | Digit DP                   | O(D * 2 * 10)   | O(D * 2)      |
//!
//! Where D = number of digits in N.

use std::cmp::min;

// =============================================================================
// TSP Bitmask — Traveling Salesman Problem
// =============================================================================

/// Solves the Traveling Salesman Problem using bitmask DP.
///
/// # Problem
/// Given a complete weighted graph with `n` cities, find the minimum cost
/// to visit every city exactly once and return to the starting city.
///
/// # State
/// `dp[mask][i]` = minimum cost to reach city `i` having visited exactly
/// the set of cities represented by `mask`, starting from city 0.
///
/// - `mask` is a bitmask: bit `j` is set if city `j` has been visited.
/// - Transition: `dp[mask | (1 << j)][j] = min(dp[mask][i] + dist[i][j])`
///   for each unvisited city `j` adjacent to `i`.
///
/// # Complexity
/// - Time: O(n^2 * 2^n) — for each of 2^n masks and n cities, try n transitions
/// - Space: O(n * 2^n) — the DP table
///
/// # Practical Limit
/// Feasible for n <= ~20 due to exponential state space.
fn tsp_bitmask(dist: &[Vec<i64>]) -> i64 {
    let n = dist.len();
    let full_mask = (1 << n) - 1;
    let inf = i64::MAX / 2;

    // dp[mask][i] = min cost to reach city i with visited set = mask
    let mut dp = vec![vec![inf; n]; 1 << n];
    dp[1][0] = 0; // Start at city 0, only city 0 visited (mask = 0b...001)

    for mask in 1..(1 << n) {
        for u in 0..n {
            if dp[mask][u] >= inf {
                continue;
            }
            // Bit u must be set in mask (we are at city u)
            if mask & (1 << u) == 0 {
                continue;
            }

            // Try visiting each unvisited city
            for v in 0..n {
                if mask & (1 << v) != 0 {
                    continue; // Already visited
                }
                let new_mask = mask | (1 << v);
                let new_cost = dp[mask][u] + dist[u][v];
                if new_cost < dp[new_mask][v] {
                    dp[new_mask][v] = new_cost;
                }
            }
        }
    }

    // Find minimum cost to complete the tour (return to city 0)
    let mut result = inf;
    for u in 0..n {
        if dp[full_mask][u] < inf && dist[u][0] < inf {
            result = min(result, dp[full_mask][u] + dist[u][0]);
        }
    }

    result
}

// =============================================================================
// LIS Optimal — O(n log n) Longest Increasing Subsequence
// =============================================================================

/// Finds the length of the longest strictly increasing subsequence using
/// the patience sorting algorithm.
///
/// # Algorithm
/// Maintain a vector `tails` where `tails[i]` = smallest tail element of
/// any increasing subsequence of length `i + 1` found so far.
///
/// `tails` is always sorted, enabling binary search via `partition_point`.
///
/// For each element `num`:
/// - Binary search for the leftmost position where `tails[pos] >= num`.
/// - If `pos == tails.len()`, `num` extends the longest subsequence.
/// - Otherwise, `tails[pos] = num` (replace to keep tail minimal).
///
/// # Why O(n log n)?
/// - n elements processed, each with an O(log n) binary search.
///
/// # Rust Note
/// `partition_point` is Rust's built-in binary search that finds the first
/// index where the predicate returns `false`. It's equivalent to C++'s
/// `lower_bound`.
///
/// # Complexity
/// - Time: O(n log n)
/// - Space: O(n)
fn lis_optimal(nums: &[i32]) -> usize {
    if nums.is_empty() {
        return 0;
    }

    let mut tails: Vec<i32> = Vec::new();

    for &num in nums {
        let pos = tails.partition_point(|&x| x < num);
        if pos == tails.len() {
            tails.push(num);
        } else {
            tails[pos] = num;
        }
    }

    tails.len()
}

// =============================================================================
// Matrix Chain Multiplication
// =============================================================================

/// Finds the minimum number of scalar multiplications needed to multiply
/// a chain of matrices.
///
/// # Problem
/// Given matrices A1 (p0 x p1), A2 (p1 x p2), ..., An (p(n-1) x pn),
/// find the optimal parenthesization that minimizes total multiplications.
///
/// # Recurrence
/// `dp[i][j]` = minimum multiplications to compute matrices i through j.
/// ```text
/// dp[i][j] = min over k in [i, j-1] of:
///     dp[i][k] + dp[k+1][j] + dims[i-1] * dims[k] * dims[j]
/// ```
///
/// # Complexity
/// - Time: O(n^3) — three nested loops
/// - Space: O(n^2)
///
/// # Parameters
/// - `dims`: array of dimensions where matrix i has dimensions
///   `dims[i-1] x dims[i]`. Length is n+1 for n matrices.
fn matrix_chain_multiplication(dims: &[i64]) -> i64 {
    let n = dims.len() - 1; // Number of matrices
    if n <= 1 {
        return 0;
    }

    // dp[i][j] = min cost to multiply matrices i..j (1-indexed)
    let mut dp = vec![vec![0i64; n + 1]; n + 1];

    // l = chain length (2 to n)
    for l in 2..=n {
        for i in 1..=n - l + 1 {
            let j = i + l - 1;
            dp[i][j] = i64::MAX;

            // Try all possible split points
            for k in i..j {
                let cost = dp[i][k] + dp[k + 1][j] + dims[i - 1] * dims[k] * dims[j];
                if cost < dp[i][j] {
                    dp[i][j] = cost;
                }
            }
        }
    }

    dp[1][n]
}

// =============================================================================
// Digit DP — Count numbers in [1, N] that don't contain digit 4
// =============================================================================

/// Counts how many numbers in the range [1, N] do NOT contain the digit 4.
///
/// # Digit DP Framework
/// Process the number digit by digit (most significant to least significant).
/// At each position, maintain:
/// - `tight`: whether we're still bounded by the original number N
///   (if true, the next digit can be at most `digits[pos]`; if false,
///   any digit 0-9 is allowed)
///
/// # State
/// `dp[pos][tight]` = count of valid numbers formed by digits at positions
/// `pos..end`, given the `tight` constraint.
///
/// # Complexity
/// - Time: O(D * 2 * 10) where D = number of digits in N
/// - Space: O(D * 2) for memoization
///
/// # Counting Trick
/// We compute count(0..=N) and subtract 1 to exclude 0, giving count(1..=N).
fn digit_dp_no_four(n: u64) -> u64 {
    if n == 0 {
        return 0;
    }

    let digits: Vec<u32> = n.to_string()
        .chars()
        .map(|c| c.to_digit(10).unwrap())
        .collect();

    let len = digits.len();

    // Memoization: memo[pos][tight]
    // -1 means not computed yet
    let mut memo = vec![vec![-1i64; 2]; len];

    fn solve(
        pos: usize,
        tight: bool,
        digits: &[u32],
        memo: &mut Vec<Vec<i64>>,
    ) -> i64 {
        if pos == digits.len() {
            return 1; // Successfully formed a valid number
        }

        let tight_idx = if tight { 1 } else { 0 };
        if memo[pos][tight_idx] != -1 {
            return memo[pos][tight_idx];
        }

        let upper = if tight { digits[pos] } else { 9 };
        let mut count = 0i64;

        for d in 0..=upper {
            if d == 4 {
                continue; // Skip digit 4
            }
            count += solve(
                pos + 1,
                tight && d == upper,
                digits,
                memo,
            );
        }

        memo[pos][tight_idx] = count;
        count
    }

    let total = solve(0, true, &digits, &mut memo);
    // Subtract 1 to exclude the number 0 (which has no digit 4 but isn't in [1, N])
    (total - 1) as u64
}

// =============================================================================
// Main — Test cases
// =============================================================================

fn main() {
    println!("=== Week 23: Advanced DP ===\n");

    // --- TSP Bitmask ---
    println!("--- TSP Bitmask ---");
    let dist = vec![
        vec![0, 10, 15, 20],
        vec![10, 0, 35, 25],
        vec![15, 35, 0, 30],
        vec![20, 25, 30, 0],
    ];
    let result = tsp_bitmask(&dist);
    println!("TSP minimum tour cost: {} (expected 80)", result);
    assert_eq!(result, 80); // 0->1->3->2->0: 10+25+30+15=80

    // Smaller test
    let dist2 = vec![
        vec![0, 1, 2],
        vec![1, 0, 3],
        vec![2, 3, 0],
    ];
    let result2 = tsp_bitmask(&dist2);
    println!("TSP (3 cities): {} (expected 6)", result2);
    assert_eq!(result2, 6); // 0->1->2->0: 1+3+2=6
    println!("PASS\n");

    // --- LIS Optimal ---
    println!("--- LIS Optimal (O(n log n)) ---");
    let lis = lis_optimal(&[10, 9, 2, 5, 3, 7, 101, 18]);
    println!("LIS([10,9,2,5,3,7,101,18]) = {} (expected 4)", lis);
    assert_eq!(lis, 4);

    let lis2 = lis_optimal(&[0, 1, 0, 3, 2, 3]);
    assert_eq!(lis2, 4);

    let lis3 = lis_optimal(&[7, 7, 7, 7]);
    assert_eq!(lis3, 1);

    let lis4 = lis_optimal(&[1, 2, 3, 4, 5]);
    assert_eq!(lis4, 5); // Already sorted

    let lis5 = lis_optimal(&[5, 4, 3, 2, 1]);
    assert_eq!(lis5, 1); // Reverse sorted
    println!("PASS\n");

    // --- Matrix Chain Multiplication ---
    println!("--- Matrix Chain Multiplication ---");
    // Matrices: A1 (10x30), A2 (30x5), A3 (5x60)
    let dims = vec![10, 30, 5, 60];
    let result = matrix_chain_multiplication(&dims);
    println!("MCM([10,30,5,60]) = {} (expected 4500)", result);
    assert_eq!(result, 4500); // (A1*A2)*A3: 10*30*5 + 10*5*60 = 1500+3000

    // 4 matrices: A1 (40x20), A2 (20x30), A3 (30x10), A4 (10x30)
    let dims2 = vec![40, 20, 30, 10, 30];
    let result2 = matrix_chain_multiplication(&dims2);
    println!("MCM([40,20,30,10,30]) = {} (expected 26000)", result2);
    assert_eq!(result2, 26000);
    println!("PASS\n");

    // --- Digit DP ---
    println!("--- Digit DP (no digit 4) ---");
    let count = digit_dp_no_four(10);
    println!("Numbers in [1,10] without digit 4: {} (expected 9)", count);
    // 1,2,3,5,6,7,8,9,10 — only 4 is excluded
    assert_eq!(count, 9);

    let count2 = digit_dp_no_four(100);
    println!("Numbers in [1,100] without digit 4: {} (expected 81)", count2);
    // 1-digit: 8, 2-digit (10-99): 8*9=72, plus 100 -> 81
    assert_eq!(count2, 81);

    let count3 = digit_dp_no_four(4);
    println!("Numbers in [1,4] without digit 4: {} (expected 3)", count3);
    assert_eq!(count3, 3); // 1, 2, 3

    let count4 = digit_dp_no_four(44);
    println!("Numbers in [1,44] without digit 4: {}", count4);
    // 1-digit: 8 (1,2,3,5,6,7,8,9)
    // 2-digit 10-39: first digit {1,2,3} (3 choices), second {0-9}\{4} (9) = 27
    // 40-44: first digit 4 -> skip all
    // Total = 8 + 27 = 35
    assert_eq!(count4, 35);
    println!("PASS\n");

    println!("All Week 23 tests passed!");
}

File: Week 23/cpp/advanced_dp.cpp
/*
 * =============================================================================
 * Week 23 — Advanced Dynamic Programming  (C++17)
 * =============================================================================
 *
 * Topics covered
 * --------------
 *   1. TSP — Travelling Salesman Problem (bitmask DP)
 *   2. Max independent set on a tree (tree DP)
 *   3. Digit DP — count numbers up to N with specific digit properties
 *   4. LIS O(n log n) — patience sorting with std::lower_bound
 *   5. Matrix Chain Multiplication
 *
 * Complexity cheat-sheet
 * ----------------------
 *   tsp_bitmask              O(n^2 * 2^n)   |  Space O(n * 2^n)
 *   max_independent_set      O(V + E)        |  Space O(V)
 *   digit_dp                 O(D * sum * 2)  |  Space O(D * sum * 2)
 *   lis_nlogn                O(n log n)      |  Space O(n)
 *   matrix_chain_mult        O(n^3)          |  Space O(n^2)
 *
 * Build & run
 *   g++ -std=c++17 -O2 -o advanced_dp advanced_dp.cpp && ./advanced_dp
 * =============================================================================
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>
#include <sstream>
#include <climits>
#include <cstring>
#include <functional>
#include <string>
#include <numeric>

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------
template <typename T>
std::string vec_str(const std::vector<T>& v) {
    std::ostringstream oss;
    oss << "[";
    for (std::size_t i = 0; i < v.size(); ++i) {
        if (i) oss << ", ";
        oss << v[i];
    }
    oss << "]";
    return oss.str();
}

// ---------------------------------------------------------------------------
// 1. TSP — bitmask DP
// ---------------------------------------------------------------------------
// dp[mask][i] = minimum cost to visit all cities in `mask`, ending at city i.
// Start and end at city 0.
// Complexity:  Time O(n^2 * 2^n)  |  Space O(n * 2^n)
int tsp_bitmask(const std::vector<std::vector<int>>& dist) {
    int n = static_cast<int>(dist.size());
    int full = (1 << n) - 1;
    std::vector<std::vector<int>> dp(1 << n, std::vector<int>(n, INT_MAX));

    dp[1][0] = 0;  // start at city 0

    for (int mask = 1; mask <= full; ++mask) {
        for (int u = 0; u < n; ++u) {
            if (dp[mask][u] == INT_MAX) continue;
            if (!(mask & (1 << u))) continue;

            for (int v = 0; v < n; ++v) {
                if (mask & (1 << v)) continue;  // already visited
                int new_mask = mask | (1 << v);
                dp[new_mask][v] = std::min(dp[new_mask][v], dp[mask][u] + dist[u][v]);
            }
        }
    }

    // Return to city 0.
    int ans = INT_MAX;
    for (int u = 1; u < n; ++u) {
        if (dp[full][u] != INT_MAX) {
            ans = std::min(ans, dp[full][u] + dist[u][0]);
        }
    }
    return ans;
}

// ---------------------------------------------------------------------------
// 2. Max Independent Set on a Tree
// ---------------------------------------------------------------------------
// dp[v][0] = max weight independent set in subtree of v, NOT including v
// dp[v][1] = max weight independent set in subtree of v, INCLUDING v
// Complexity:  Time O(V)  |  Space O(V)
int max_independent_set(const std::vector<std::vector<int>>& adj,
                        const std::vector<int>& weights) {
    int n = static_cast<int>(adj.size());
    if (n == 0) return 0;

    std::vector<std::array<int, 2>> dp(n, {0, 0});
    std::vector<bool> visited(n, false);

    std::function<void(int)> dfs = [&](int u) {
        visited[u] = true;
        dp[u][1] = weights[u];

        for (int v : adj[u]) {
            if (visited[v]) continue;
            dfs(v);
            dp[u][0] += std::max(dp[v][0], dp[v][1]);  // u not taken, v can be either
            dp[u][1] += dp[v][0];                        // u taken, v must not be taken
        }
    };

    dfs(0);  // root at 0
    return std::max(dp[0][0], dp[0][1]);
}

// ---------------------------------------------------------------------------
// 3. Digit DP — count integers in [1, N] whose digit sum equals S
// ---------------------------------------------------------------------------
// Complexity:  Time O(D * S * 2)  |  Space O(D * S * 2)
//              D = number of digits in N, S = target sum
long long digit_dp_count(const std::string& num_str, int target_sum) {
    int D = static_cast<int>(num_str.size());
    if (target_sum < 0) return 0;

    // dp[pos][sum][tight]
    // Memoize with -1 as unvisited.
    std::vector<std::vector<std::vector<long long>>> memo(
        D, std::vector<std::vector<long long>>(target_sum + 1, std::vector<long long>(2, -1)));

    std::function<long long(int, int, bool)> solve = [&](int pos, int sum, bool tight) -> long long {
        if (sum > target_sum) return 0;
        if (pos == D) return (sum == target_sum) ? 1 : 0;
        if (memo[pos][sum][tight] != -1) return memo[pos][sum][tight];

        int limit = tight ? (num_str[pos] - '0') : 9;
        long long result = 0;

        for (int d = 0; d <= limit; ++d) {
            result += solve(pos + 1, sum + d, tight && (d == limit));
        }

        return memo[pos][sum][tight] = result;
    };

    // This counts in [0, N] with digit sum = target_sum.
    // Subtract 1 if target_sum == 0 (to exclude 0 itself from [1, N]).
    long long ans = solve(0, 0, true);
    if (target_sum == 0) ans -= 1;
    return ans;
}

// ---------------------------------------------------------------------------
// 4. LIS — O(n log n) with lower_bound (patience sorting)
// ---------------------------------------------------------------------------
// tails[i] = smallest tail of all increasing subsequences of length i+1
// Complexity:  Time O(n log n)  |  Space O(n)
int lis_nlogn(const std::vector<int>& nums) {
    std::vector<int> tails;

    for (int x : nums) {
        auto it = std::lower_bound(tails.begin(), tails.end(), x);
        if (it == tails.end()) {
            tails.push_back(x);
        } else {
            *it = x;
        }
    }
    return static_cast<int>(tails.size());
}

// ---------------------------------------------------------------------------
// 5. Matrix Chain Multiplication
// ---------------------------------------------------------------------------
// Given dimensions p[0..n], matrix i has dimensions p[i] x p[i+1].
// Find minimum scalar multiplications needed.
// dp[i][j] = min cost to multiply matrices i..j
// Complexity:  Time O(n^3)  |  Space O(n^2)
long long matrix_chain_mult(const std::vector<int>& p) {
    int n = static_cast<int>(p.size()) - 1;  // number of matrices
    if (n <= 1) return 0;

    std::vector<std::vector<long long>> dp(n, std::vector<long long>(n, 0));

    // l = chain length
    for (int l = 2; l <= n; ++l) {
        for (int i = 0; i <= n - l; ++i) {
            int j = i + l - 1;
            dp[i][j] = LLONG_MAX;
            for (int k = i; k < j; ++k) {
                long long cost = dp[i][k] + dp[k + 1][j]
                               + (long long)p[i] * p[k + 1] * p[j + 1];
                dp[i][j] = std::min(dp[i][j], cost);
            }
        }
    }
    return dp[0][n - 1];
}

// ---------------------------------------------------------------------------
// main — test cases
// ---------------------------------------------------------------------------
int main() {
    std::cout << "=== Week 23: Advanced DP ===\n\n";

    // 1. TSP bitmask DP
    {
        std::cout << "-- TSP (bitmask DP) --\n";
        std::vector<std::vector<int>> dist = {
            {0, 10, 15, 20},
            {10,  0, 35, 25},
            {15, 35,  0, 30},
            {20, 25, 30,  0}
        };
        int ans = tsp_bitmask(dist);
        std::cout << "  min tour cost: " << ans << "\n";
        assert(ans == 80);  // 0->1(10)->3(25)->2(30)->0(15) = 80
        std::cout << "\n";
    }

    // 2. Max Independent Set on Tree
    {
        std::cout << "-- Max Independent Set (Tree DP) --\n";
        //       0(w=1)
        //      / \
        //    1(10) 2(5)
        //    /
        //  3(8)
        std::vector<std::vector<int>> adj = {
            {1, 2},   // 0's neighbors
            {0, 3},   // 1's neighbors
            {0},      // 2's neighbors
            {1}       // 3's neighbors
        };
        std::vector<int> weights = {1, 10, 5, 8};
        int mis = max_independent_set(adj, weights);
        std::cout << "  max independent set weight: " << mis << "\n";
        // Take nodes 1 and 2: 10+5=15  (can't take 0 with 1 or 2; can't take 3 with 1)
        // Or take 0, 3: 1+8=9
        // Or take 2, 3: 5+8=13
        // Best: take 1 and 2 = 15? But 1 and 2 are both children of 0, they are independent.
        assert(mis == 15);
        std::cout << "\n";
    }

    // 3. Digit DP
    {
        std::cout << "-- Digit DP --\n";
        // Count integers in [1, 100] with digit sum = 5
        // 5, 14, 23, 32, 41, 50 => 6 numbers
        long long cnt = digit_dp_count("100", 5);
        assert(cnt == 6);
        std::cout << "  numbers in [1,100] with digit sum 5: " << cnt << "\n";

        // Count in [1, 20] with digit sum = 3: 3, 12 => 2
        long long cnt2 = digit_dp_count("20", 3);
        assert(cnt2 == 2);
        std::cout << "  numbers in [1,20] with digit sum 3:  " << cnt2 << "\n\n";
    }

    // 4. LIS O(n log n)
    {
        std::cout << "-- LIS O(n log n) --\n";
        std::vector<int> nums = {10, 9, 2, 5, 3, 7, 101, 18};
        int len = lis_nlogn(nums);
        assert(len == 4);
        std::cout << "  nums=" << vec_str(nums) << "  LIS=" << len << "\n";

        std::vector<int> nums2 = {0, 1, 0, 3, 2, 3};
        int len2 = lis_nlogn(nums2);
        assert(len2 == 4);  // 0,1,2,3
        std::cout << "  nums=" << vec_str(nums2) << "  LIS=" << len2 << "\n\n";
    }

    // 5. Matrix Chain Multiplication
    {
        std::cout << "-- Matrix Chain Multiplication --\n";
        // Matrices: A1(10x30), A2(30x5), A3(5x60)
        // Optimal: (A1*A2)*A3 = 10*30*5 + 10*5*60 = 1500 + 3000 = 4500
        std::vector<int> p = {10, 30, 5, 60};
        auto cost = matrix_chain_mult(p);
        assert(cost == 4500);
        std::cout << "  dims=" << vec_str(p) << "  min cost=" << cost << "\n";

        // Matrices: A1(40x20), A2(20x30), A3(30x10), A4(10x30)
        std::vector<int> p2 = {40, 20, 30, 10, 30};
        auto cost2 = matrix_chain_mult(p2);
        assert(cost2 == 26000);
        std::cout << "  dims=" << vec_str(p2) << "  min cost=" << cost2 << "\n";
    }

    std::cout << "\nAll Week 23 tests passed.\n";
    return 0;
}
